### DISCUSSION TITLE
Correct Grouping of related Entities

### DISCUSSION THREAD

**Discussion by stsiushkevich** [feature-sliced/documentation#632] [Q&A] [answered]
Hello! I have a typical situation when there are some set of related entities(slices). For example we have ```Employee``` entity, and related ```EmployeeGroup``` and ```EmployeeRoleGroup``` entities and appropriate TS Types. For all of them there are according data lists, that displayed on the page and fetching data from the server. Should I leave these slices like separated at the same ```@entities``` layer, or unite them to a single slice ```Employee```?

![image](https://github.com/feature-sliced/documentation/assets/12151060/799c949e-8b94-4a87-8534-5afb2b165703)

**Comment by illright** [comment, marked as answer]
I would recommend keeping the slices flat as long as possible, and only when it starts to overflow with slices, you can start grouping them. Note that when you group slices into a folder, they don't become one slice. In the case below:
* entities/
  * Employee/
    * Person/
    * Group/
    * RoleGroup/
  * Identification/

The slices are `Employee/Person`, `Employee/Group`, and `Employee/RoleGroup`, and these slices are bound by the same rules as `Identification`, no code sharing is permitted between `Employee/*` slices.

**Reply by stsiushkevich** [reply]
Ok, but can the following structure be alternatively approach that you described:
```@entities/Employee/api/services/EmployeeService```
```@entities/Employee/api/services/EmployeeGroupService```
```@entities/Employee/api/services/EmployeeRoleGroupService```

```@entities/Employee/ui/EmployeeList```
```@entities/Employee/ui/EmployeeGroupList```
```@entities/Employee/ui/EmployeeRoleGroupList```

...... etc.......

So here we have only one single common/collective slice ```Employee```, but many entities in it. Or maybe it is bad approach because it can seem that we can use entity ```EmployeeList``` from ```EmployeeGroupList``` for example?

**Reply by illright** [reply]
This is a perfectly valid approach too. Usually you should split an entity into several entities (potentially grouped) when the single entity becomes unmaintainable. Most decisions in FSD are ruled by the principle of **Pain-Driven Development**: do something only when not doing it causes you pain. If keeping it all inside the `Employee` entity is fine, do that.

> it is bad approach because it can seem that we can use entity EmployeeList from EmployeeGroupList for example
For the reasons described above, it's not bad. The rules of slice isolation are meant to provoke thinking, but in some cases no thinking is required. It's not automatically bad when one _business entity_ uses another, but only as long as they are contained within a single slice (so from the perspective of FSD, they are all just one entity). It depends on how deeply you want to decompose.

**Reply by stsiushkevich** [reply]
Thank you a lot!

**Reply by dante01yoon** [reply]
> I would recommend keeping the slices flat as long as possible, and only when it starts to overflow with slices, you can start grouping them. Note that when you group slices into a folder, they don't become one slice. In the case below:
> 
> * entities/
>   
>   * Employee/
>     
>     * Person/
>     * Group/
>     * RoleGroup/
>   * Identification/
> 
> The slices are `Employee/Person`, `Employee/Group`, and `Employee/RoleGroup`, and these slices are bound by the same rules as `Identification`, no code sharing is permitted between `Employee/*` slices.

Hi @illright, 
isn't it `Employee` is slice and each nested  Person, Group, RoleGoup is  `segments`?
Also, isn't it available sharing code between segment is possible?

**Reply by illright** [reply]
Hi @dante01yoon, no, in this case, Employee is a slice group. Slice groups are just regular folders with slices. Slices in the same group are bound by the same rules as other slices — the slice `Employee/Person` cannot share any code with `Employee/Group`. Notice that when I write the names of these slices, I include the group in the name.

**Reply by dante01yoon** [reply]
@illright Hey thanks for details.  How deeply can I manage slice directory ?
in this case, you suggested `Employee/Person` as a slice group, can I make it deep nested like `Employee/Someting/Person` or more than?

**Reply by illright** [reply]
You can go as deep as you want, but I'd advise to take caution here. Usually, too deep of a grouping is a sign of overslicing

**Reply by dante01yoon** [reply]
Got it, thanks !
